### ISSUE TITLE
Running tailscale on CircleCI fails with SO_BINDTODEVICE

### ISSUE THREAD

**Issue by chrisgoffinet** [tailscale/tailscale#2827]
I am attempting to use tailscale 1.14.0 with userspace-networking on CircleCI. It looks to be failing with `setting SO_BINDTODEVICE: operation not permitted`, I made sure to run as `root`. 

Here is the log:
```
$ sudo tailscaled -verbose 10 --tun=userspace-networking
logtail started
Program starting: v1.14.0-t5cff36945-g809e87bba, Go 1.17-tsdd6b269c77: []string{"tailscaled", "-verbose", "10", "--tun=userspace-networking"}
LogID: 7226a55e02478af0f1ea71db909049bc7042675461dba456521d594577bd3870
logpolicy: using system state directory "/var/lib/tailscale"
logpolicy.Read /var/lib/tailscale/tailscaled.log.conf: open /var/lib/tailscale/tailscaled.log.conf: no such file or directory
wgengine.NewUserspaceEngine(tun "userspace-networking") ...
using fake (no-op) tun device
using fake (no-op) OS network configurator
using fake (no-op) DNS configurator
dns: using dns.noopManager
link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[192.168.96.3/20]} v4=true v6=false}
magicsock: unable to bind udp4 port 0: listen udp4 :0: setting SO_BINDTODEVICE: operation not permitted
wgengine.NewUserspaceEngine(tun "userspace-networking") error: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
wgengine.New: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
flushing log.
logger closing down
logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
logtail: upload: log upload of 369 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/38ca472036f521dff487f563bdbb5282e18fed63ef2620c93c820b10596b1d4b": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
logtail: backoff: 12 msec
logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
logtail: upload: log upload of 369 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/38ca472036f521dff487f563bdbb5282e18fed63ef2620c93c820b10596b1d4b": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
logtail: backoff: 23 msec
logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
logtail: upload: log upload of 369 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/38ca472036f521dff487f563bdbb5282e18fed63ef2620c93c820b10596b1d4b": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
logtail: backoff: 95 msec
logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
logtail: upload: log upload of 369 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/38ca472036f521dff487f563bdbb5282e18fed63ef2620c93c820b10596b1d4b": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
logtail: backoff: 227 msec
logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
logtail: upload: log upload of 369 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/38ca472036f521dff487f563bdbb5282e18fed63ef2620c93c820b10596b1d4b": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
logtail: backoff: 132 msec
logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 4ms)
logtail: upload: log upload of 369 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/38ca472036f521dff487f563bdbb5282e18fed63ef2620c93c820b10596b1d4b": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
logtail: backoff: 511 msec
```

Is there anything we can do to make it work? Happy to share more data from the debuggable SSH session I have for the job.

**Comment by DentonGentry** [comment]
That happens in https://github.com/tailscale/tailscale/blob/main/net/netns/netns_linux.go

It is important that sockets which will be used to contact Tailscale infrastructure, like the coordination server or DERP relays, use the underlying Internet connection to do so not try to go over tailscale0. Trying to use tailscale to run tailscale might sortof work, but is mostly unable to recover from any error and never works well.

netns will first try using FW marks (requires CAP_NET_ADMIN) and falls back to SO_BINDTODEVICE (requires CAP_NET_RAW). Without either one, it is stuck.

Most likely CircleCI is running the container without CAP_NET_RAW. If there is a way to let it have CAP_NET_RAW or CAP_NET_ADMIN, it would resolve the issue.

**Comment by DentonGentry** [comment]
We're going to relax the need for CAP_NET_RAW. SO_BINDTODEVICE is needed because we don't want Tailscale control packets to egress from the `tailscale0` interface. However in `--tun=userspace-networking` mode, there is no `tailscale0` interface. There is no reason to bind to a device, so we'll just skip doing so.

A package on the development track https://pkgs.tailscale.com/unstable/ *after* 1.15.105 will do this, which you could try. If it works, it would appear in the 1.16.0 release.

**Comment by DentonGentry** [comment]
I think this should be fixed in 1.15.111, please reopen the bug if not.
https://pkgs.tailscale.com/unstable/

**Comment by chrisgoffinet** [comment]
Still seeing errors.

```
2021/09/10 17:44:38 logtail started
2021/09/10 17:44:38 Program starting: v1.15.111-t382b349c5-gcff6d9ac3, Go 1.17-tsdd6b269c77: []string{"tailscaled", "--tun=userspace-networking", "--socks5-server=localhost:1055"}
2021/09/10 17:44:38 LogID: f842751491943708a5c50530a115dc65f651a0766ed2edb4f5b6fdf408d2f8e6
2021/09/10 17:44:38 logpolicy: using system state directory "/var/lib/tailscale"
logpolicy.Read /var/lib/tailscale/tailscaled.log.conf: open /var/lib/tailscale/tailscaled.log.conf: no such file or directory
2021/09/10 17:44:38 wgengine.NewUserspaceEngine(tun "userspace-networking") ...
2021/09/10 17:44:38 dns: using dns.noopManager
2021/09/10 17:44:38 link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[192.168.176.3/20]} v4=true v6=false}
2021/09/10 17:44:38 magicsock: unable to bind udp4 port 0: listen udp4 :0: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:38 wgengine.NewUserspaceEngine(tun "userspace-networking") error: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
2021/09/10 17:44:38 wgengine.New: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
2021/09/10 17:44:38 flushing log.
2021/09/10 17:44:38 logger closing down
2021/09/10 17:44:38 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 3ms)
2021/09/10 17:44:38 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:38 logtail: backoff: 9 msec
2021/09/10 17:44:38 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 2ms)
2021/09/10 17:44:38 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:38 logtail: backoff: 55 msec
2021/09/10 17:44:38 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
2021/09/10 17:44:38 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:38 logtail: backoff: 86 msec
2021/09/10 17:44:38 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
2021/09/10 17:44:38 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:38 logtail: backoff: 184 msec
2021/09/10 17:44:39 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
2021/09/10 17:44:39 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:39 logtail: backoff: 203 msec
2021/09/10 17:44:39 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
2021/09/10 17:44:39 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:39 logtail: backoff: 400 msec
2021/09/10 17:44:39 logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted (in 1ms)
2021/09/10 17:44:39 logtail: upload: log upload of 92 bytes not-compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/28bb0c23f821dfc61d63517fda8b351e0f808d1a7bef0ff2cacddd82ab74f950": dial tcp 34.210.105.16:443: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:39 logtail: backoff: 300 msec
```

**Comment by DentonGentry** [comment]
Something clearly went wrong, it shouldn't have called SO_BINDTODEVICE.

```
2021/09/10 17:44:38 Program starting: v1.15.111-t382b349c5-gcff6d9ac3, Go 1.17-tsdd6b269c77: []string{"tailscaled", "--tun=userspace-networking", "--socks5-server=localhost:1055"}
...
2021/09/10 17:44:38 magicsock: unable to bind udp4 port 0: listen udp4 :0: setting SO_BINDTODEVICE: operation not permitted
2021/09/10 17:44:38 wgengine.NewUserspaceEngine(tun "userspace-networking") error: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
```

**Comment by chrisgoffinet** [comment]
Looks like the issue is in `tailscaled.go`

https://github.com/tailscale/tailscale/blob/382b349c54ed38bbc4d2936a8c9de6cc2f905eb8/cmd/tailscaled/tailscaled.go#L295

We need to be disabling netns before we call `createEngine()` for userspace-networking. I did my own build doing that, and it works.

/cc @bradfitz

**Comment by DentonGentry** [comment]
Would you like to send a pull request?

**Comment by bradfitz** [comment]
@chrisgoffinet, can you try out https://github.com/tailscale/tailscale/pull/2832 ?

**Comment by chrisgoffinet** [comment]
@bradfitz tested, works! we can close this out.

**Comment by DentonGentry** [comment]
If you would be willing to write up something about how you're using Tailscale with Circle CI, like as a blog post on your own site, we'd love to highlight it in the newsletter.

**Comment by chrisgoffinet** [comment]
> If you would be willing to write up something about how you're using Tailscale with Circle CI, like as a blog post on your own site, we'd love to highlight it in the newsletter.

absolutely.

**Comment by chrisgoffinet** [comment]
@DentonGentry do you happen to know when a new unstable build will go out with this latest edition, I'll need to update the orb

**Comment by DentonGentry** [comment]
We released 1.15.116 about an hour ago, which would contain the fix.

**Comment by chrisgoffinet** [comment]
@DentonGentry here is the blog post for the newsletter https://blog.threecomma.io/circleci-meet-tailscale
